### ISSUE TITLE
Breaks with astropy 4.0dev

### ISSUE THREAD

**Issue by aarchiba** [nanograv/PINT#550]
I think the problem has to do with different handling of units? Specifically I think units pass through more functions than they used to - I saw np.interp(number, number, number*u.us)*u.us have a unit of u.us2. It seems like *everything* breaks.

**Comment by aarchiba** [comment]
The following CHANGES.rst entry explains:
```
- For numpy 1.17 and later, the new ``__array_function__`` protocol is used to
  ensure that all top-level numpy functions interact properly with
  ``Quantity``, preserving units also in operations like ``np.concatenate``.
  [#8808]
```
This means we have to go through and do a careful units audit. Which will be good for us.

**Comment by paulray** [comment]
This sounds difficult to do while maintaining backwards compatibility.

**Comment by aarchiba** [comment]
Actually the Right Way is what we should have been doing all along: never pass Quantities to functions that are not explicitly Quantity-aware. The code that breaks assumes that (for example) np.interp *strips* units from its inputs, which it does in pre-4, but in 4 it just passes them through. If we just convert to the unit we want, take value, call the function, and multiply by the resulting unit, we're sorted. If the function was right before, it's just a matter of sticking .value on the end of its arguments; if the code didn't explicitly request particular units before calling such a function, it probably gave the wrong answer when presented with milliseconds instead of seconds (say).

When I talk about a careful units audit, I mean that anywhere there is a .value not immediately preceded by a .to(), there's a large chance there is a lurking bug just waiting to bite users. Someone (me I guess) should go through systematically and fix all of those.

Such an audit might not be necessary for the 4.0 update because almost any code that relies on pure numpy stripping units will immediately cause a unit conversion problem. This means about 2/3 of our tests fail, but possibly with the addition of some actual unit (rather than integration) tests it should be pretty easy to track down the problems, and very easy to tell when they're all gone.

**Comment by paulray** [comment]
I don't quite get why that is the "Right Way".  What makes a function explicitly Quantity-aware?
If functions are migrating towards all supporting quantities, shouldn't we go that direction too, and not strip and re-add units with every function call?

I guess it would be nicer if there were a way to ask a numpy/scipy/astropy function if it is Quantity-aware and only do the .value hack if it is not.

**Comment by scottransom** [comment]
I agree with @paulray   That is not the right way at all.  It is *certainly* not how the Astropy units maintainers want you to write code.  Please do *not* change all the code to do this.  We should talk to Martin van Kerkwijk about this.

**Comment by luojing1211** [comment]
I agree to have a unit optimization. A lot of our astropy.units usages are not optimized. For this specific case, I can ask Marten about it.

**Comment by mhvk** [comment]
Yes, the idea is that from astropy 4.0/numpy 1.17 onward all (well, almost all) numpy functions will be unit-aware and "do the right thing" - i.e., it does the `.value` and stick the unit back on for you. I do realize this is a huge pain for supporting both older and newer astropy/numpy, but am obviously hoping that long-term this will make really help, with units no longer disappearing.

My guess, though, is that when you look more carefully, you'll find it is not that many places where you go wrong, and that often it will be quite clear whether one should just to `.to_value(<some_unit>)` or use the take-out-and-stick-back-on method. Old/new could could be separated using the following (future-proved?) way:
```
try:
    # If this module exists, numpy functions may be overridden by Quantity
    from astropy.units.quantity_helper import function_helpers
    # In future versions, this attribute may disappear as it is always true.
    ARRAY_FUNCTION_ENABLED = getattr(function_helpers, 'ARRAY_FUNCTION_ENABLED', True)
except ImportError:
    # Older astropy.
    ARRAY_FUNCTION_ENABLED = False
```

Of course, you could also decide this is the time to make a clean break - python>=3.7, astropy>=4.0, numpy>=1.17.........

**Comment by aarchiba** [comment]
> Of course, you could also decide this is the time to make a clean break - python>=3.7, astropy>=4.0, numpy>=1.17.........

This is the sticking point: we are stuck supporting python2 for, I'd say, at least another year, maybe more. Which pins us to astropy 2.0.x. So we can't assume any numpy function will pass units through. 

Thus our choices are, for every call to a function we don't trust to handle Quantities:

1. strip units on input, re-add them after, or
2. wrap it in an if statement that does 1 for old astropy and leaves the units intact for new astropy, or 
3. wrap it in an if statement that uses the current code to strip units and then re-apply them and leaves the units as-is for new astropy. 

We therefore need to go do *something* for every non-Quantity-aware function we call with Quantities. So for every occurrence:

- Do we want to add an if statement?
- Do we want to touch the old version? If the old version is correct, to turn 3 into 1 we would just need to replace .to(unit) with .to(unit).value . If the old version does not to .do(unit) it's probably wrong and we need to touch it anyway.

**Comment by aarchiba** [comment]
> I don't quite get why that is the "Right Way". What makes a function explicitly Quantity-aware?
> If functions are migrating towards all supporting quantities, shouldn't we go that direction too, and not strip and re-add units with every function call?

I would love to be in a position to trust all the functions we use to behave correctly, or at least raise an exception, when handed a Quantity. But we have to support python2, where almost all compiled functions in numpy will silently strip units that are provided to them.

> I guess it would be nicer if there were a way to ask a numpy/scipy/astropy function if it is Quantity-aware and only do the .value hack if it is not.

The easiest way for us in this situation is to just switch to astropy 4 and then go hunt down the problems that make 2/3 of our test suite fail. Almost all of those are due to us assuming that a numpy function would strip units but now it doesn't any more. All of those need to be changed to work with astropy 4; if we want them to work on both 4 and pre-4 either we write an if statement every time or we feed them unitless quantities and apply the units after the fact. This is what we currently do for most of the numpy functions that *don't* break when switched to astropy 4; it was the correct thing to do in pre-4.

For packages that are not written with astropy in mind, there are about four possibilities for any given function:

1. It was written in python in a way that is Quantity-safe. When passed Quantity inputs, it will raise an exception if they are incompatible and produce a result with the correct units if they are compatible.
2. It was written in python in a way that implodes when passed a Quantity. This can happen in model-fitting and smoothing functions, for example, where coefficients are fit for that have to have all sorts of different units to make their combinations with different powers and combinations of the inputs combinable. This will usually but not always raise an exception when passed a Quantity. 
3. It was written in C with no special attention. This will silently strip units from all its inputs and produce a non-Quantity numpy array as output.
4. It was written in C with appropriately magical use of array subclasses. I'm not totally sure this is possible but functions designed to preserve matrix-ness might preserve units from their input on the output. This might even be the correct units. (Not if the function is "reciprocal" of course.)

These have always been the limitations of Quantity; numpy 1.17 allows astropy 4.0 to make some compiled functions behave like category 1. Unfortunately many places in our code we assume these functions are in category 3.

**Comment by mhvk** [comment]
It may be best to first find out how many places actually cause problems; my suspicion is it will be not be that many, in which case `if/else` constructs are not too hard (and they certainly make your future self very happy, as stripping out the old cruft will be easy). But my experience may not be typical (e.g., my code probably uses quite a bit less scipy than PINT).

Aside: can one not just go the bug-fix only route for PINT on python2? Or is really the problem that people need latest and greatest PINT but cannot upgrade to python3? I still find that surprising. And it does come at a cost of repelling at least some potential contributors (well, at least me).

**Comment by scottransom** [comment]
I think that we could go bugfix-only for python2 basically now, actually.  I don't know of anyone who *must* have python2 capability.

**Comment by aarchiba** [comment]
> I think that we could go bugfix-only for python2 basically now, actually. I don't know of anyone who _must_ have python2 capability.

That doesn't actually help very much, as we're so rife with bugs. Almost everything I've done has been fixing bugs, variously defined. That would also mean maintaining two different versions. I'm not sure we have the sustained developer weight to maintain *one* version; we haven't ever had a release of any kind, and we have quite a few stale old bugs and PRs that have been sitting in GitHub for months or years. I think keeping the maintenance burden low would be really valuable.

There are a few situations where I would love to use modern python features - if we're going to drop old-python support I'd love to see us depend on >=3.6 so we can use f-strings, for example. On the other hand, if the entire Fermi toolchain is python2 only, dropping python2 support makes PINT vastly less useful for working with Fermi data. I just don't think it's worth it.

**Comment by paulray** [comment]
My vote is that we stick with astropy<4.0 for a little while, at least until 4.x is stable and has had time to propagate into the usual distribution channels (pip, MacPorts, conda).  Then we can make a release and declare it the last release that supports Python 2.7.  At that point, we can change to developing only for Python>=3.6, astropy>=4.0 and remove the current unit hacks for functions that strip the units. I think we can declare that we won't be doing much to fix bugs in the Python 2.7 version.  Hopefully PRESTO will be at 3.x by then and the fermitools are supposed to get there by the end of the year as well.

**Comment by scottransom** [comment]
@aarchiba We don't *need* to be bug-free (who is?).  We don't *need* to have releases.  We don't *need* to support everyone.  We don't *need* to have perfect test coverage.  We don't *need* to have everything perfect before we publish a paper.  We have gotten very good results out of PINT so far (and TEMPO, TEMPO2, PRESTO, PSRCHIVE, ...) without any of those things.

That being said, I *love* that you are pushing for all of this stuff.  And I *really* appreciate it.  Thank you!  It is really helping us do things better and hopefully will bring in more users and developers in the future.  But we don't want the burdens to be too high.

As for the Fermi toolchain, for instance, that is a completely self-contained thing.  I don't think PINT calls anything (or at least not much) from it.  And even if it does, that is a very specific use case for PINT.  We can deal with that (for example, with a broken sub-package for a while).  And people can do their Fermi processing in 2.7 and then switch to 3.6+ for their PINT work.  @paulray Do you actively use the Fermi toolchain and PINT in the same code?  And if so, do you know if that is a common thing?  I suspect it is not.

I'm not sure why you don't think we couldn't go to python 3.6+ right now if that is the best way to deal with the numpy/astropy changes.  We are all moving past 2.7 for the *vast* majority of uses within the next 6 months to a year anyways (including PRESTO very soon -- I'm getting so sick of myself not having time to perfectly finish things, I might just complete the pull request and deal with bug reports in the short term.)

**Comment by paulray** [comment]
Indeed, the Fermi tools are totally independent of PINT (and vice versa). I use the Fermi-installed version of Python for them and that is not what I use for PINT. So I agree we definitely don't need to be on the same version of Python as the Fermi tools.

**Comment by paulray** [comment]
Also note that we do have releases (depending on how you define "release").  We have tagged versions, as described here:
https://github.com/nanograv/PINT/wiki/Tagging-PINT-Versions

Of course, we haven't been disciplined about using them or following semantic versioning yet (given our 0.x status), but the intent was there.

**Comment by mhvk** [comment]
p.s. astropy 4.0 will be a long-term support version.

**Comment by aarchiba** [comment]
Part of our problem turns out to be https://github.com/astropy/astropy/issues/9357 which is as far as I can tell a plain bug in astropy 4.0dev

**Comment by matteobachetti** [comment]
@aarchiba, @paulray, @scottransom: just to let you know that now that Astropy 4 is the default "stable" (and not dev anymore!), all packages depending on PINT have this bug. In Stingray, we have to stick to Astropy 3 for now if we want to use the Pulsar machinery.

**Comment by matteobachetti** [comment]
And yet another numpy/astropy change biting: https://github.com/astropy/astropy/issues/8841

(See failures at: https://travis-ci.org/StingraySoftware/stingray/jobs/631799275?utm_medium=notification&utm_source=github_status)

**Comment by paulray** [comment]
We do support astropy 4, so I think this can be closed (although 4.0.1 is a different story, see #651).
